game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java

```diff
@@ -142,20 +148,55 @@ public Image getImage(final UnitCategory unit) {
       final String baseImageName, final PlayerId id, final ResourceLoader resourceLoader) {
     // URL uses '/' not '\'
     final String fileName = FILE_NAME_BASE + id.getName() + "/" + baseImageName + ".png";
-    final URL url = resourceLoader.getResource(fileName);
+    final String fileName2 = FILE_NAME_BASE + baseImageName + ".png";
+    final URL url = resourceLoader.getResource(fileName, fileName2);
     return Optional.ofNullable(url);
   }
 
-  private Optional<Image> getBaseImage(final String baseImageName, final PlayerId id) {
+  private Optional<Image> getBaseImage(
+      final String baseImageName, final PlayerId id, final UnitType type) {
     final Optional<URL> imageLocation = getBaseImageUrl(baseImageName, id);
     Image image = null;
     if (imageLocation.isPresent()) {
       image = Toolkit.getDefaultToolkit().getImage(getBaseImageUrl(baseImageName, id).get());
       Util.ensureImageLoaded(image);
+      if (!mapData.ignoreColorizingUnit(type.getName())) {
```

Comment:
IMO `UnitImageFactory` is mainly about finding and loading images from disk.

The fact it has unit highlight is a bit of a code smell where it's no longer really even a factory but is also a  `UnitImageRenderer`.

It would be nice to have a full-blown adherence to the 'open-closed' principle be applied to this class, but it would be a pretty heavy lift. In the meantime, I also note we have a `mapData` dependency injected into this class, a very heavy dependency. Notably, [interface segregation principle](https://en.wikipedia.org/wiki/Interface_segregation_principle) comes to mind.

Just as the logic for unit colorization is externalized, I think the image rendering logic could also be externalized to, or moved to the colorizer class. In which case, the code here would be something like:

```
private final UnitColorTransform unitColorTransformer;
private final ImageOrientationTransform orientationTransform

image = unitColorTransform.apply(image);
image = orientationTransform.apply(image);
```

With the above, if the map data flag is not set, then we'd have a no-op and just return the parameter value. The above decreases the cyclomatic complexity and keeps UnitImageFactory from knowing everything about how images are drawn.

What do you think?

Reply:
Probably a valid point and I decided to at least move the image flipping outside of this. This could probably be taken further but would probably be better to look at the class as a whole since several other methods like highlight already do even more image manipulation then this new functionality.

Reply:
That modularity does not strike me as correct. Why not have the image manipulation also be in colorizer, why make unit image factory do that? It seems we're building on problems. I really see problems with introducing mapData as a dependency. That would torpedo testing, violates interface segregation principle, and if we build on it will take large effect to undo. I don't think modifying existing code to this extent to add in case statements is the way to go. Easier to avoid this now. I do think it's simple to extract this new code to a new class, instead of injecting map data, inject the new class.

Reply:
I don't want to push TripleA specific implemention into `ImageTransformer` as its meant to be a generic image manipulation class that could even one day be extracted into a separate library if it became useful outside of this project. Really the only image logic left here is checking all the map data properties to determine if and what to transform. The transformation itself is in `ImageTransformer`.

Where that logic belongs you could debate but given how the classes are structured currently probably fits best where it is unless a new class is introduced. Particularly this chunk of code:
```
        if (mapData.getUnitColor(id.getName()).isPresent()) {
          final Color color = mapData.getUnitColor(id.getName()).get();
          final int brightness = mapData.getUnitBrightness(id.getName());
          ImageTransformer.colorize(color, brightness, (BufferedImage) image);
        }
        if (mapData.shouldFlipUnit(id.getName())) {
          image = ImageTransformer.flipHorizontally((BufferedImage) image);
        }
      }
```

Reply:
> or create a new class for these 20 or so lines

I don't have any problem with a 20 line class file. It encapsulates the image rendering code and buffers UnitImageFactory from knowing about MapData. I suspect such a class might be readily testable, which is a good sign for design.

IMO it's similar to how the raw colorizer code is extracted so we don't have UnitImageFactory dealing with the raw colorizer directly nor map data. I think that creates a pretty nice design tree where we have map data and colorizer coming together in one module (class), and then unit image factory uses that class.

Reply:
Resolving this comment so we can just have the conversation in one comment thread that I responded to.